src/apps/chifra/pkg/output/stream.go

```
@@ -0,0 +1,177 @@
+	// This Flushes for each printed item, but in the exchange the user gets
+	// the data printed as it comes
+	outputWriter.Flush()
+
+	err := outputWriter.Error()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// StreamRaw outputs raw `Raw` to `w`
+func StreamRaw[Raw types.RawData](w io.Writer, raw *Raw) (err error) {
+	bytes, err := json.MarshalIndent(raw, "", "  ")
+	if err != nil {
+		return
+	}
+	w.Write(bytes)
+	// Add a newline so that the command prompt is not being printed at
+	// the same line as the output
+	w.Write([]byte("\n"))
+	return
+}
+
+// StreamMany outputs models or raw data as they are acquired
+func StreamMany[Raw types.RawData](
+	w io.Writer,
+	getData func(models chan types.Modeler[Raw], errors chan error),
+	options OutputOptions,
+) error {
+	models := make(chan types.Modeler[Raw])
+	errors := make(chan error)
+
+	// Check if the current item is the first that we print. If so, we may want to
+	// print keys or postpone adding JSON comma between the elements
+	first := true
+	// Start getting the data
+	go func() {
+		getData(models, errors)
+		close(models)
+		close(errors)
+	}()
+
+	// If we are printing JSON, we want to make sure that opening and closing
+	// brackets are printed
+	if options.Format == "json" {
+		w.Write([]byte("{\n  \"data\": [\n    "))
+		defer w.Write([]byte("\n  ]\n}\n"))
+	}
+	// If printing API format, we want to add meta information
+	if options.Format == "api" {
+		w.Write([]byte("{\n  \"data\": [\n    "))
+		defer func() {
+			w.Write([]byte("\n  ], \n\"meta\": "))
+			b, _ := json.MarshalIndent(options.Meta, "", options.JsonIndent)
+			w.Write(b)
+			w.Write([]byte("\n}\n"))
+		}()
+	}
+
+	// If user wants custom format, we have to prepare the template
+	customFormat := strings.Contains(options.Format, "{")
+	tmpl, err := template.New("").Parse(options.Format)
+	if customFormat && err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case model, ok := <-models:
+			if !ok {
+				continue
```

Will this ever happen? If not, can we panic here instead of silently ignore?